### PR TITLE
Variable type traits to check whether a variable is continuous or categorical

### PR DESCRIPTION
**Pull request by andrei-sa**

### CODE REVIEW

File: Python/Include/VariableTraits.hpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    }
+    {
+      auto val = obj < obj;
+    }
+    {
+      auto val = S() * obj;
+    }
+    try {
+      auto val = obj / obj;
+    } catch(pybind11::error_already_set& e) {
```
Comment (kranar) [inline Python/Include/VariableTraits.hpp]: Catch by ```const &```.

File: Python/Include/VariableTraits.hpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      e.restore();
+      if(PyErr_ExceptionMatches(PyExc_ZeroDivisionError)) {
+        PyErr_Clear();
+      } else {
+        throw pybind11::error_already_set();
```
Comment (kranar) [inline Python/Include/VariableTraits.hpp]: Use ```throw;``` to rethrow the exception.

File: Python/Include/VariableTraits.hpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  struct ContinuousChecker<S, pybind11::object> {
+
+    //! Checks whether a python object is continuous for modeling purposes for
+    //! scalar S.
+    static bool check(const pybind11::object& obj);
+  };
+
+  template<typename S>
+  bool ContinuousChecker<S, pybind11::object>::check(const pybind11::object&
+      obj) try {
```
Comment (kranar) [inline Python/Include/VariableTraits.hpp]: Function try-blocks are only used in constructors and only when absolutely needed.

File: Python/Include/VariableTraits.hpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    {
+      auto val = obj + obj;
```
Comment (kranar) [inline Python/Include/VariableTraits.hpp]: Omit the assignment and introduction of the new scope:

```
  obj + obj;
  -obj;
  obj == obj;
  obj != obj;
  ...
```

**Review by kranar — CHANGES_REQUESTED** [review]
Good stuff.

File: Python/Include/VariableTraits.hpp (hunk 1 above)
Comment (andrei-sa) [reply Python/Include/VariableTraits.hpp]: `error_already_set::restore()` is non-const. It releases the exception to Python C API, which seems necessary to check for the exception type (`ZeroDivisionError` in this case).

File: Python/Include/VariableTraits.hpp (hunk 2 above)
Comment (andrei-sa) [reply Python/Include/VariableTraits.hpp]: `error_already_set` constructor fetches the exception from Python C API, which has intercepted the original exception due to `restore()`, so it is the only way to correctly rethrow the context of the original exception. In this case though, it does not really matter as the exception is not processed.

Comment (kranar) [reply Python/Include/VariableTraits.hpp]: Okay thanks.

File: Python/Include/VariableTraits.hpp (hunk 1 above)
Comment (kranar) [reply Python/Include/VariableTraits.hpp]: Yep that's right.

File: Python/Include/VariableTraits.hpp [hunk 5]
```diff
@@ -21,38 +21,28 @@ namespace Rover {
...
+      obj < obj;
+      S() * obj;
+      try {
+        obj / obj;
+      } catch(pybind11::error_already_set& e) {
+        e.restore();
+        if(PyErr_ExceptionMatches(PyExc_ZeroDivisionError)) {
+          PyErr_Clear();
+        } else {
+          throw pybind11::error_already_set();
```
Comment (kranar) [inline Python/Include/VariableTraits.hpp]: Could you change this to ```return false;```?

**Review by kranar — CHANGES_REQUESTED** [review]

**Review by kranar — APPROVED** [review]
This is great stuff.

**Pull request merged**